include/pybind11/numpy.h

```diff
@@ -1093,21 +1093,33 @@ class array_t : public array {
     // Reference to element at a given index
     template <typename... Ix>
     const T &at(Ix... index) const {
-        if ((ssize_t) sizeof...(index) != ndim()) {
-            fail_dim_check(sizeof...(index), "index dimension mismatch");
-        }
-        return *(static_cast<const T *>(array::data())
-                 + byte_offset(ssize_t(index)...) / itemsize());
+        check_dim_precondition(sizeof...(index));
+        return const_reference(index...);
     }
 
     // Mutable reference to element at a given index
     template <typename... Ix>
     T &mutable_at(Ix... index) {
-        if ((ssize_t) sizeof...(index) != ndim()) {
-            fail_dim_check(sizeof...(index), "index dimension mismatch");
-        }
-        return *(static_cast<T *>(array::mutable_data())
-                 + byte_offset(ssize_t(index)...) / itemsize());
+        check_dim_precondition(sizeof...(index));
+        return mutable_reference(index...);
+    }
+
+    // const-reference to element at a given index without bounds checking
+    template <typename... Ix>
+    const T &operator()(Ix... index) const {
+#if !defined(NDEBUG)
```

Comment:
IMO it's an important feature, a test for it would be ideal (so that we don't accidentally lose it), and I think it's easy:

In tests/test_numpy_array.cpp:

```
    m.attr("defined_NDEBUG") =
#ifdef NDEBUG
    true;
#else
    false;
#endif
```

Then in tests/test_numpy_array.py replace `test_at_fail()` with (untested):

```
@pytest.mark.parametrize(
    "func",
    [
        m.at_t,
        m.mutate_at_t,
        m.const_subscript_via_call_operator_t,
        m.subscript_via_call_operator_t,
    ][: 2 if m.defined_NDEBUG else 99],
)
@pytest.mark.parametrize("dim", [0, 1, 3])
def test_elem_reference(arr, func, dim):
    with pytest.raises(IndexError) as excinfo:
        func(arr, *([0] * dim))
    assert str(excinfo.value) == f"index dimension mismatch: {dim} (ndim = 2)"
```

Reply:
yes thank you for suggesting this - I was not super familiar with the pybind11 testing so this is helpful for me to understand how things should be done